spring/src/main/java/io/cucumber/spring/SpringFactory.java

```diff
@@ -139,8 +139,10 @@ public void start() {
         notifyContextManagerAboutTestClassStarted();
         if (beanFactory == null || isNewContextCreated()) {
             beanFactory = testContextManager.getBeanFactory();
-            for (Class<?> stepClass : stepClasses) {
-                registerStepClassBeanDefinition(beanFactory, stepClass);
+            synchronized (beanFactory) {
+                for (Class<?> stepClass : stepClasses) {
+                    registerStepClassBeanDefinition(beanFactory, stepClass);
```

Comment:
registerStepClassBeanDefinition is not threadsafe. Code in DefaultListableBeanFactory

```java
            if (this.hasBeanCreationStarted()) {
                synchronized(this.beanDefinitionMap) { //its synchronized, but it doesnt check again before adding to the updatedDefinitions list
                    this.beanDefinitionMap.put(beanName, beanDefinition);
                    List<String> updatedDefinitions = new ArrayList(this.beanDefinitionNames.size() + 1);
                    updatedDefinitions.addAll(this.beanDefinitionNames);
                    updatedDefinitions.add(beanName);
                    this.beanDefinitionNames = updatedDefinitions;
```